gcloud-java-core/src/main/java/com/google/gcloud/BaseServiceException.java

```diff
@@ -16,26 +16,80 @@
 
 package com.google.gcloud;
 
+import com.google.api.client.googleapis.json.GoogleJsonError;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+
+import java.io.IOException;
+import java.net.SocketTimeoutException;
+import java.util.Collections;
+import java.util.Set;
+
 /**
  * Base class for all service exceptions.
  */
 public class BaseServiceException extends RuntimeException {
 
   private static final long serialVersionUID = 5028833760039966178L;
+  public static final int UNKNOWN_CODE = 0;
 
   private final int code;
   private final boolean retryable;
+  private final boolean idempotent;
+
+  public BaseServiceException(IOException exception, boolean idempotent) {
+    super(exception.getMessage(), exception);
+    this.code = statusCode(exception);
+    this.idempotent = idempotent;
+    this.retryable = idempotent && isRetryable(exception);
+  }
+
+  public BaseServiceException(GoogleJsonError error, boolean idempotent) {
+    super(error.getMessage());
+    this.code = error.getCode();
+    this.idempotent = idempotent;
+    this.retryable = idempotent && isRetryable(error);
+  }
 
-  public BaseServiceException(int code, String message, boolean retryable) {
+  public BaseServiceException(int code, String message, boolean retryable, boolean idempotent) {
     super(message);
     this.code = code;
-    this.retryable = retryable;
+    this.idempotent = idempotent;
+    this.retryable = idempotent && retryable;
   }
 
-  public BaseServiceException(int code, String message, boolean retryable, Exception cause) {
+  public BaseServiceException(int code, String message, boolean retryable, boolean idempotent,
+      Exception cause) {
     super(message, cause);
     this.code = code;
-    this.retryable = retryable;
+    this.idempotent = idempotent;
+    this.retryable = idempotent && retryable;
+  }
+
+  protected Set<Integer> retryableCodes() {
+    return Collections.emptySet();
+  }
+
+  protected int statusCode(IOException exception) {
+    if (exception instanceof GoogleJsonResponseException) {
```

Comment:
I think the changes you made to DatastoreException and DatastoreRpcException are reasonable however I wonder what do you think if we apply these changes instead:

1) include a notion of a reason (String) to base service exception using `GoogleJsonError.getMessage()`.

2) replace `retryableCodes` with a pair of (code, reason).

3) retryable is true only when there is a match of the pair, however null value means *any*.
  - it should be possible to have e.g. (null, "InternalError") [null code but with a reason] to
    support the `StorageException` case.

4) change DatastoreRpcException to throw GoogleJsonResponseException (translate `com.google.api.services.datastore.client.DatastoreException` to `GoogleJsonResponseException` in `DefaultDatastoreRpc` and change signature of the DatastoreRpc to throw IOExcetion instead of DatastoreRpcException.


* I would still hold off in changing the `Datastore` api layer to throw `IllegalArgumentException` instead of `DatastoreException` and do it in a separate PR (or change the api javadoc in this one).